### ISSUE TITLE
Rails 5 schema.rb adding non-agnostic options.  dev mysql -> test sqlite

### ISSUE THREAD

**Issue by duffyjp** [rails/rails#26209]
### Steps to reproduce

Use MySQL for Development, and SQLite3 for Test.  On an app I'm upgrading from 4.2 to 5.0 I run into trouble **after** running a migration in the Rails5 branch. Here's why:

Rails 4:

``` bash
ack "create_table" db/schema.rb 
  create_table "absences", force: :cascade do |t|
  create_table "accounts", force: :cascade do |t|
  create_table "addresses", force: :cascade do |t|
  ...
```

Rails 5:

``` bash
ack "create_table" db/schema.rb 
  create_table "absences", force: :cascade, options: "ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci" do |t|
  create_table "accounts", force: :cascade, options: "ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci" do |t|
  create_table "addresses", force: :cascade, options: "ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci" do |t|
  ...
```

That's fine for MySQL, but when I load the schema for tests I get this:

```
Tasks: TOP => db:test:load_schema
(See full trace by running task with --trace)
ActiveRecord::StatementInvalid: SQLite3::SQLException: near "ENGINE": syntax error: CREATE TABLE "absences" ("id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, "person_id" integer, "date" date, "source" varchar, "appointment_id" varchar, "reason" varchar, "category" varchar, "hours" decimal(4,2), "approved" boolean, "created_at" datetime NOT NULL, "updated_at" datetime NOT NULL, "entry_source" varchar, "voided" boolean DEFAULT 'f' NOT NULL, "mark_for_deletion" boolean DEFAULT 'f') ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci
```
### Expected behavior

Migration ignores options it's can't use, or they're not in the schema.rb in the first place.
### Actual behavior

sqlite3 (1.3.11) builds invalid SQL (per sqlite3).
### System configuration

Working: 4.2.5.1
Broken: 5.0.0.1

**Ruby version**:
2.3.1p112

**Comment by vipulnsward** [comment]
Can you share your database config wrt extra options?

**Comment by duffyjp** [comment]
It's pretty stock.  Mysql for the local DB, and sqlite3 for test.  

```
development:
  adapter: mysql2
  encoding: utf8
  database: xxxx
  pool: 32
  username: xxxx
  password: xxxx
  host: localhost

test:
  adapter: sqlite3
  database: db/test.sqlite3<%= ENV['TEST_ENV_NUMBER'] %>
  pool: 5
  timeout: 5000
```

**Comment by duffyjp** [comment]
I setup a blank Rails 5.0.0.1 app, added mysql2, and set development to mysql and test/prod to sqlite3.  

Add a migration:
```ruby
rails g migration create_foos bar
```


### Observations of schema.rb


rake db:setup
```ruby
ActiveRecord::Schema.define(version: 20160822152937) do

  create_table "foos", force: :cascade do |t|
    t.string "bar"
  end

end

```

rake db:migrate
```ruby
ActiveRecord::Schema.define(version: 20160822152937) do

  create_table "foos", force: :cascade, options: "ENGINE=InnoDB DEFAULT CHARSET=utf8" do |t|
    t.string "bar"
  end

end

```




database.yml
```
default: &default
  adapter: sqlite3
  pool: 5
  timeout: 5000

development:
  adapter: mysql2
  encoding: utf8
  database: issue26209
  username: root
  password: your_pass_here
  host: 127.0.0.1

test:
  <<: *default
  database: db/test.sqlite3

production:
  <<: *default
  database: db/production.sqlite3
```

**Comment by sgrif** [comment]
This is the expected behavior. The role of `schema.rb` is to provide a format which is easier for diffs and less prone to merge conflicts. The addition of those options is necessary for properly reproducible builds. We strongly recommend against using a different database for development and test mode.

**Comment by duffyjp** [comment]
@sgrif Could the new behavior be made optional?  This is a practice I've used since Rails 3 beta without issue.  Using sqlite for testing is also very common on Travis CI, where folks will use in-memory databases, or with [parallel_tests](https://github.com/grosser/parallel_tests) for the ease of creating numerous temporary database copies.  I've spoken to a number of developers who use this approach despite using MySQL or Postgres in production.  I find a real value in testing against a different database, as it forces you to write database agnostic code.

I understand it may be discouraged, but I respectfully disagree it should be prevented.

**Comment by sgrif** [comment]
If a config option were to be added, it wouldn't be until 5.1 at the earliest, but yes if you are interested in working on a pull request I would merge one which added a configuration option to not dump table options in `schema.rb`.

**Comment by duffyjp** [comment]
@sgrif The documentation states the schema file is database-independent.

http://edgeguides.rubyonrails.org/active_record_migrations.html#what-are-schema-files-for-questionmark

> In many ways this is exactly what it is. This file is created by inspecting the database and expressing its structure using create_table, add_index, and so on. Because this is database-independent, it could be loaded into any database that Active Record supports. This could be very useful if you were to distribute an application that is able to run against multiple databases.
> 
> There is however a trade-off: db/schema.rb cannot express database specific items such as triggers, stored procedures or check constraints. While in a migration you can execute custom SQL statements, the schema dumper cannot reconstitute those statements from the database. If you are using features like this, then you should set the schema format to :sql.

**Comment by sgrif** [comment]
Yeah, that's a bit out of date. That hasn't really been true since 4.0

**Comment by duffyjp** [comment]
Would you be willing to reopen this issue?  There is either a bug in the code or a bug in the documentation.  No one will see this closed.  

I _really_ strongly disagree that despite ActiveRecord migrations being agnostic, the schema shouldn't be.  If the additional options in the schema are useful, they should be prefixed with the database type they apply to, and ignored if loading the schema against a different database type.

**Comment by BenFenner** [comment]
I too would like the schema to remain optionally agnostic.

Regardless, this ticket shouldn't be closed until the schema option is implemented, or at the very least until a link is provided to the new ticket that describes updating the previously mentioned out-of-date documentation. "Working as intended" while the official documentation says otherwise is not valid.

**Comment by jmuheim** [comment]
I ran into this exact same problem, too. I'd like to have SQLite as test database, and MySQL for development.

**Comment by jeremy** [comment]
To make the schema agnostic, we could qualify these table options with the database they apply to.

For example, 
```ruby
  create_table "foos", force: :cascade, mysql_options: "ENGINE=InnoDB DEFAULT CHARSET=utf8" do |t|
    t.string "bar"
  end
```

They'd simply be ignored (at your peril, if they're meaningful or necessary) on other databases.

**Comment by gusridd** [comment]
I ran into the same problem 5 minutes ago and I totally agree with @jeremy. That should solve the merge problem and allow users to use SQLite and MySQL without hassle.

**Comment by jeremy** [comment]
Let's see a PR! 😁

**Comment by lloeki** [comment]
> This is the expected behavior. The role of schema.rb is to provide a format which is easier for diffs and less prone to merge conflicts. The addition of those options is necessary for properly reproducible builds. We strongly recommend against using a different database for development and test mode.

What about those who build an app that has to support multiple database types and uses `schema.rb` to bootstrap a fresh install for each new tenant? Our tests run on both MariaDB and PostgreSQL and we are using both in development for obvious reasons. ActiveRecord is supposed to be an abstraction, not ignoring options that obviously make no sense and basically passing them straight to SQL is as leaky as it can get.

```ruby
create_table "foos", force: :cascade, options: "ENGINE=InnoDB DEFAULT CHARSET=utf8"
```

should really be:

```ruby
create_table "foos", force: :cascade, options: { mysql2: "ENGINE=InnoDB DEFAULT CHARSET=utf8" }
```

**Comment by jeremy** [comment]
@lloeki Definitely! That nicely summarizes the discussion above. Need a champion to step up with a PR.

**Comment by jmuheim** [comment]
@lloeki Exactly my opinion.

**Comment by Yanchek99** [comment]
I also would like the schema to remain optionally agnostic.  @sgrif Could we reopen this ticket as it breaks existing flexibility and there is no mention of this change in any upgrade notes.

**Comment by jordanlance** [comment]
Ran into this also, very bad. 
It should just allow us to run migrations and be coherent to the 'outdated' documentation posted above that states you can use any database.

**Comment by kamipo** [comment]
We need to keep DB agnostic `schema.rb` as possible.

**Comment by matthewd** [comment]
To copy over a comment from #29472:

Per https://github.com/rails/rails/pull/27981#issuecomment-279427794, I don't think we should ignore settings that mean something, and we should instead aim to exclude default settings from the schema dump.

**Comment by wnewbery** [comment]
I ran into this every time I wanted to use or support multiple databases (e.g. SQLite3 for some devs), and there is a lot more inconsistencies than just these options.

e.g. looking at the diff of what PostgreSQL to SQLite gave us for the exact same migrations (which do not contain any special SQL at all!):

 * SQLite3 used `integer` columns for my foreign keys (`t.belongs_to` etc.), but PostgreSQL used `bigint`.
 * SQLite3 driver doesn't support foreign keys, so all the `add_foreign_key`'s at the end got deleted.
 * Many columns were reordered.


I have not looked in depth, but with `schema.rb` being made by dumping the database (rather than concatenating migrations in some way) I am inclined to think that there could be a whole list of edge cases between the different databases where datatypes and features differ.



If Rails wants `schema.rb` to be agnostic, maybe it should instead build itself from the migrations directly? Even if DB specific stuff is to be moved to its own parameters, like `options: {mysql: ..., sqlite: ...}` we would still then want it to "merge" the dumps rather than rewrite the `schema.db` so that it contains both at the same time, not just the one the developer that committed it was using (and especially not having stuff like it deleting all the foreign keys).

Else it seems its better to say it will never truly be, and generate separate files (`schema_mysql.rb`, `schema_postgresql.rb`, `schema_sqlite.rb`, etc.) that while following the same basic format and being possible (maybe with manual effort) to move correctly between DB's, are in no way advertised as such and can co-exist in version control.

**Comment by Pluto1010** [comment]
@kamipo +1 for your fix. 

Can I assume to find that in 5.1.5 ?

**Comment by lloeki** [comment]
@wnewbery Similarly, there are differences on `text` types with `limit`: some MySQL builds may output `4294967295`, some `2147483647`, while PostgreSQL doesn't output anything, so loading that "database agnostic" schema with MySQL makes it [fall back to 65535](https://dev.mysql.com/doc/refman/5.7/en/storage-requirements.html#data-types-storage-reqs-strings), which can result in silent truncation. Introspecting schema solely from the database cannot produce a database agnostic `schema.rb` at the semantic level.

**Comment by wnewbery** [comment]
Well like I said, i dont think the current method can ever work without catches. Seperate schema file names might be better, at least if someone must rename the file they can expect issues, or rerun all migrations and have both schemas checked in. 

Different builds of the same version is harder. If you have a schema with 4294967295 and import it into a build with 2147483647 for example.

**Comment by tilo** [comment]
@sgrif what is the status of this issue? 

It is disheartening to see this issue has been raised in Aug 2018, and it has still not been addressed..

Is there an ETA when this will be fixed, and what version of Rails is targeted for the fix?


We too use SQLite3 for much faster testing, and MySQL in prod.

One other issue I noticed is that some of the limits from the old schema file no longer show up in the new format.

**Comment by tilo** [comment]
maybe @kamipo can help?

**Comment by sgrif** [comment]
@tilo Please stop pinging various team members. It will not bring more attention to your issue. As has already been stated, we're happy to accept a fix for this, but this issue needs a champion to write up a PR. It is not currently a priority for anybody on the team. Continuously pinging us here and on other mediums will not make something happen faster.

**Comment by tilo** [comment]
It's duffyjp's issue really.

So the question is _how_ one is supposed to find a "champion"?
Filing an issue and waiting for 16 months, hoping someone stumbles over it, doesn't seem to work

Introducing this sort of dependency in the schema dump seems pretty short-sighted IMHO.
AR is supposedly an abstraction, so people can easily switch out their DB..

**Comment by tilo** [comment]
@sgrif 
So there is already a PR for this, from the original author who made the change in the schema dumping, but it is still not merged:

https://github.com/rails/rails/pull/29472

**Comment by wnewbery** [comment]
@sgrif Is a complete design known, or does other things need to happen first?

For example in tilo's case, even with the PR, if a developer was to use SQLite locally rather than MySQL to add a migration and commit the resulting (SQlite specific) schema, using that later for MySQL will still have lost the new MySQL specific options, and potentially worse foreign keys and other items.

**Comment by tilo** [comment]
Sorry, I don't mean to be critical, but maybe this whole idea of saving DB-specific options was not such a good idea to begin with, given that AR is supposed to abstract from the underlying DB..?
What was the original issue this new dump format was supposed to solve? 
Is it really worth doing?

**Comment by georgeclaghorn** [comment]
The core team has been pretty clear about what needs to happen here. Further commentary is unnecessary.
